### DISCUSSION TITLE
How do I allow the use of colons (:) for shortUrls?

### DISCUSSION THREAD

**Discussion by vzrenggamani** [YOURLS/YOURLS#3777] [Q&A] [answered]
I want to use a colon as the form of the shortURL I want to use? But YOURLS doesn't seem to allow it (sanitised?).

Is there any way to bypass the colon?
Thank you very much.

References:
1. https://en.wikipedia.org/wiki/Uniform_Resource_Name
2. https://datatracker.ietf.org/doc/html/rfc3986
3. https://stackoverflow.com/questions/2053132/is-a-colon-safe-for-friendly-url-use

**Comment by dgw** [comment]
Make a copy of [the sample `hyphens-in-urls` plugin](https://github.com/YOURLS/YOURLS/tree/master/user/plugins/hyphens-in-urls) and modify it to add colons into the allowed charset.

**Reply by vzrenggamani** [reply]
Thank you, but i encounter some "new problems here", i can't add new link with these colon, its return new error.. event the database are in clean state

![image](https://github.com/YOURLS/YOURLS/assets/14950309/977c1ee4-5620-452a-8a7e-f7accfad45b0)
![image](https://github.com/YOURLS/YOURLS/assets/14950309/25c8401f-47f1-49a4-a129-f79169f0e1cd)

**Reply by dgw** [reply]
Hmm… Let me summarize what I found in the short time I had available to debug this today.

It appears that some parts of the pre-shortening logic to validate keywords run a `yourls_sanitize_keyword()` function without the parameter to perform sanitization against the shorturl charset. Without that parameter, the function sanitizes the input keyword as if it is a URL instead—[and this code path removes colons](https://github.com/YOURLS/YOURLS/blob/97f03117d0a0036b9d0af960ed35e9d0e2e42e0d/includes/functions-formatting.php#L536).

Some simple test code run on the CLI (`php test.php`) helped me track down a possibility or two:

```php
<?php
require_once('includes/load-yourls.php');

echo( "Sanitize keyword: " . yourls_sanitize_keyword('some:colon:thing') . "\n");
echo( "Sanitize keyword(true): " . yourls_sanitize_keyword('some:colon:thing', true) . "\n");
echo( "free? " . yourls_keyword_is_free('some:colon:thing') . "\n" );
echo( "taken? " . yourls_keyword_is_taken('some:colon:thing') . "\n" );
echo( "reserved? " . yourls_keyword_is_reserved('some:colon:thing') . "\n" );
```

This outputs:

```
Sanitize keyword:
Sanitize keyword(true): some:colon:thing
free?
taken?
reserved? 1
```

I haven't traced it any further than this yet, but I _think_ you are running into a core bug, where during the check for "reserved-ness" the keyword is being incorrectly treated as a URL and gets sanitized to empty-string if it contains a colon. There are some other checks in the `yourls_keyword_is_reserved()` function, but sanitization gone wrong seems the most likely to cause this.

It'll need more investigation. I'm on a work trip this week, but hopefully @LeoColomb or @ozh will have time to take a look if I can't get back to it within a day or two.

**Reply by vzrenggamani** [reply]
Things got complicated by then :), so no "temp" fix for this issues yet. Thank for the time to investigate this issues..

**Reply by HughP** [reply]
I would be keen to see this fixed in core. we are using colons in URN type urls. This is a blocker for us. I couldn't get the suggested plugin method to work either. @dgw

**Reply by ozh** [reply, marked as answer]
Try this plugin : https://gist.github.com/ozh/ff7c454ce3fa9ce8ffeaba07f8324caa
